File: src/Parse/Citations_reg.php
<?php

namespace WpRefs\Parse\Reg_Citations;

/*
Usage:

use function WpRefs\Parse\Reg_Citations\get_name;
use function WpRefs\Parse\Reg_Citations\get_Reg_Citations;
use function WpRefs\Parse\Reg_Citations\get_full_refs;
use function WpRefs\Parse\Reg_Citations\getShortCitations;

*/

// include_once __DIR__ . '/../WikiParse/Citations_reg.php';

/**
 * Get the name attribute from citation options.
 *
 * @param string $options The citation options string to extract name from
 * @return string The extracted name or empty string if not found
 */

function get_name($options)
{
    if (trim($options) == "") {
        return "";
    }
    // $pa = "/name\s*=\s*\"(.*?)\"/i";
    $pa = "/name\s*\=\s*[\"\']*([^>\"\']*)[\"\']*\s*/iu";
    preg_match($pa, $options, $matches);
    // ---
    if (!isset($matches[1])) {
        return "";
    }
    $name = trim($matches[1]);
    return $name;
}
function get_Reg_Citations($text)
{
    preg_match_all("/<ref([^\/>]*?)>(.+?)<\/ref>/isu", $text, $matches);
    // ---
    $citations = [];
    // ---
    foreach ($matches[1] as $key => $citation_options) {
        $content = $matches[2][$key];
        $ref_tag = $matches[0][$key];
        $options = $citation_options;
        $citation = [
            "content" => $content,
            "tag" => $ref_tag,
            "name" => get_name($options),
            "options" => $options
        ];
        $citations[] = $citation;
    }

    return $citations;
}

function get_full_refs($text)
{
    $full = [];
    $citations = get_Reg_Citations($text);
    // ---
    foreach ($citations as $cite) {
        $name = $cite["name"];
        $ref = $cite["tag"];
        // ---
        $full[$name] = $ref;
    };
    // ---
    return $full;
}

function getShortCitations($text)
{
    preg_match_all("/<ref ([^\/>]*?)\/\s*>/isu", $text, $matches);
    // ---
    $citations = [];
    // ---
    foreach ($matches[1] as $key => $citation_options) {
        $ref_tag = $matches[0][$key];
        $options = $citation_options;
        $citation = [
            "content" => "",
            "tag" => $ref_tag,
            "name" => get_name($options),
            "options" => $options
        ];
        $citations[] = $citation;
    }

    return $citations;
}

File: python/src/Parse/Citations_reg.py
"""
Citations regex parser module

Implemented from: src/Parse/Citations_reg.php

Usage:
    from src.Parse.Citations_reg import get_name, get_Reg_Citations, get_full_refs, getShortCitations
"""

import re


def get_name(options: str) -> str:
    """
    Get the name attribute from citation options
    
    Matches: src/Parse/Citations_reg.php - get_name()
    
    Args:
        options: The citation options string to extract name from
        
    Returns:
        The extracted name or empty string if not found
    """
    if not options or not options.strip():
        return ""
    
    # Pattern to extract name attribute
    pattern = r'name\s*=\s*["\']?([^>"\'\s]*)["\']?\s*'
    match = re.search(pattern, options, re.IGNORECASE)
    
    if not match:
        return ""
    
    name = match.group(1).strip()
    return name


def get_Reg_Citations(text: str) -> list:
    """
    Get citations using regex
    
    Matches: src/Parse/Citations_reg.php - get_Reg_Citations()
    
    Args:
        text: Text to parse
        
    Returns:
        List of citation dictionaries with keys: content, tag, name, options
    """
    pattern = r'<ref([^/>]*?)>(.+?)</ref>'
    matches = re.findall(pattern, text, re.IGNORECASE | re.DOTALL)
    full_matches = re.findall(r'<ref[^/>]*?>.+?</ref>', text, re.IGNORECASE | re.DOTALL)
    
    citations = []
    
    for i, (citation_options, content) in enumerate(matches):
        ref_tag = full_matches[i] if i < len(full_matches) else ""
        options = citation_options
        citation = {
            "content": content,
            "tag": ref_tag,
            "name": get_name(options),
            "options": options
        }
        citations.append(citation)
    
    return citations


def get_full_refs(text: str) -> dict:
    """
    Get full references from text
    
    Matches: src/Parse/Citations_reg.php - get_full_refs()
    
    Args:
        text: Text to parse
        
    Returns:
        Dictionary mapping reference names to their full tags
    """
    full = {}
    citations = get_Reg_Citations(text)
    
    for cite in citations:
        name = cite["name"]
        ref = cite["tag"]
        full[name] = ref
    
    return full


def getShortCitations(text: str) -> list:
    """
    Get short citations (self-closing ref tags) from text
    
    Matches: src/Parse/Citations_reg.php - getShortCitations()
    
    Args:
        text: Text to parse
        
    Returns:
        List of short citation dictionaries
    """
    pattern = r'<ref ([^/>]*?)/\s*>'
    matches = re.findall(pattern, text, re.IGNORECASE | re.DOTALL)
    full_matches = re.findall(r'<ref [^/>]*?/\s*>', text, re.IGNORECASE | re.DOTALL)
    
    citations = []
    
    for i, citation_options in enumerate(matches):
        ref_tag = full_matches[i] if i < len(full_matches) else ""
        options = citation_options
        citation = {
            "content": "",
            "tag": ref_tag,
            "name": get_name(options),
            "options": options
        }
        citations.append(citation)
    
    return citations
